### PR TITLE
feat(743): add type annotations in jina

### PR DESCRIPTION
**Pull request by rameshwara**
Solution to issue #743

### CODE REVIEW

**Comment by maximilianwerk** [comment]
Hey @rameshwara,
Thanks a lot for your contribution. It is really helpful.
In order to let our CI pipeline be runnable again, could you switch back the jina-hub branch to `e07232`. Since we use the jina-hub as a submodule inside jina itself, the CI pipeline must be able to pick up the hub branch (which most probably was only updated locally on yourmachine).
Cheers

**Comment by rameshwara** [comment]
Hi @maximilianwerk, thanks for the response. Apparently I am working on the jina branch and not the jina-hub branch. So I don't understand why CI pipeline was not runnable. And I am not able to find the branch e07232. Can you please provide more details about the branch to which I need to switch to?

File: jina/docker/database.py [hunk 1]
```diff
@@ -49,13 +49,18 @@ def database(self):
     def collection(self):
         return self.database[self.collection_name]
     
-    def find(self, query: dict):
+    def insert(self, document: str) -> str:
```
Comment (JoanFM) [inline jina/docker/database.py:52]: why is this added here? maybe some merge issue?

File: jina/docker/hubio.py [hunk 2]
```diff
@@ -335,8 +337,12 @@ def _write_summary_to_db(self, summary: Dict):
                 _inserted_id = db.insert(document=build_summary)
                 self.logger.debug(f'Inserted the build + push summary in db with id {_inserted_id}')
 
-    def _write_summary_to_file(self, summary: Dict):
+    def _write_summary_to_file(self, summary: Dict) -> None:
```
Comment (JoanFM) [inline jina/docker/hubio.py:340]: same here, seems like some merge issue

File: jina/drivers/__init__.py [hunk 3]
```diff
@@ -265,11 +265,11 @@ def _apply_all(self, docs: Iterable['jina_pb2.Document'], context_doc: 'jina_pb2
     def __call__(self, *args, **kwargs):
         self._traverse_apply(self.req.docs, *args, **kwargs)
 
-    def _traverse_apply(self, docs, *args, **kwargs):
+    def _traverse_apply(self, docs: List, *args, **kwargs) -> None:
```
Comment (JoanFM) [inline jina/drivers/__init__.py:268]: docs are `Iterable['jina_pb2.Document']`

File: jina/executors/clients.py [hunk 4]
```diff
@@ -125,7 +126,7 @@ def get_response(self, request: 'predict_pb2.PredictRequest'):
             raise ValueError
         return self.get_output(_response)
 
-    def get_output(self, response: grpc.UnaryUnaryMultiCallable):
+    def get_output(self, response: grpc.UnaryUnaryMultiCallable) -> None:
```
Comment (JoanFM) [inline jina/executors/clients.py:129]: not really convinced if this is true for child classes, and I think in these cases, interface type is cleaner if provided by the parent

File: jina/executors/decorators.py [hunk 5]
```diff
@@ -217,7 +217,7 @@ def arg_wrapper(self, data, label=None, *args, **kwargs):
         return _batching
 
 
-def _get_size(data: Union[Iterator[Any], List[Any], np.ndarray], axis: int = 0) -> int:
+def _get_size(data: Union[Iterator[Any], List[Any], np.ndarray], axis: int = 0) -> Union[Tuple, int, None]:
```
Comment (JoanFM) [inline jina/executors/decorators.py:220]: better to say `Optional[Tuple, int]`

File: jina/executors/indexers/keyvalue.py [hunk 6]
```diff
@@ -44,15 +44,15 @@ def query(self, key: int) -> Optional['jina_pb2.Document']:
         if self.query_handler is not None:
             return self.query_handler.get(key, None)
 
-    def get_query_handler(self):
+    def get_query_handler(self) -> None:
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:47]: since children behave differently I would not put this one

File: jina/helper.py [hunk 7]
```diff
@@ -358,7 +358,7 @@ def get_random_identity() -> str:
 yaml = _get_yaml()
 
 
-def expand_env_var(v: str) -> str:
+def expand_env_var(v: str) -> Union[str, Any]:
```
Comment (JoanFM) [inline jina/helper.py:361]: Union[str, Any] seems equivalent to `Any`?

File: jina/peapods/__init__.py [hunk 8]
```diff
@@ -12,7 +12,7 @@
     import argparse
 
 
-def Pea(args: 'argparse.Namespace' = None, allow_remote: bool = True, **kwargs):
+def Pea(args: 'argparse.Namespace' = None, allow_remote: bool = True, **kwargs) -> Union['RemotePea', 'ContainerPea', 'HeadPea', 'TailPea', 'BasePea']:
```
Comment (JoanFM) [inline jina/peapods/__init__.py:15]: `BasePea` is the parent class, so by claiming u return `BasePea` you are having a correct statement

File: jina/peapods/__init__.py [hunk 9]
```diff
@@ -47,7 +47,7 @@ def Pea(args: 'argparse.Namespace' = None, allow_remote: bool = True, **kwargs):
         return BasePea(args)
 
 
-def Pod(args: Union['argparse.Namespace', Dict] = None, allow_remote: bool = True, **kwargs):
+def Pod(args: Union['argparse.Namespace', Dict] = None, allow_remote: bool = True, **kwargs) -> Union['MutablePod', 'RemoteMutablePod', 'RemotePod', 'BasePod']:
```
Comment (JoanFM) [inline jina/peapods/__init__.py:50]: equivalent here

File: jina/peapods/pea.py [hunk 10]
```diff
@@ -56,7 +56,7 @@ def __call__(cls, *args, **kwargs) -> 'PeaMeta':
         return type.__call__(_cls, *args, **kwargs)
 
 
-def _get_event(obj: 'BasePea') -> Event:
+def _get_event(obj: 'BasePea') -> Union[Event, None]:
```
Comment (JoanFM) [inline jina/peapods/pea.py:59]: `Optional[Event]`

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by maximilianwerk** [comment]
> Hi @maximilianwerk, thanks for the response. Apparently I am working on the jina branch and not the jina-hub branch. So I don't understand why CI pipeline was not runnable. And I am not able to find the branch e07232. Can you please provide more details about the branch to which I need to switch to?

We use git submodules in order to include the jina-hub code into the jina repository. It might have happened, that you (by accident or since you worked on an older master branch) changed to an older version of this git submodule. In order to get back to the latest one, you can run `git submodule update --init` which should set it back to the currently used one. Please give it a try. Furthermore, here is an example to showcase how changing the head of the submodule would look like in the jina repository:

image

**Comment by rameshwara** [comment]
> > Hi @maximilianwerk, thanks for the response. Apparently I am working on the jina branch and not the jina-hub branch. So I don't understand why CI pipeline was not runnable. And I am not able to find the branch e07232. Can you please provide more details about the branch to which I need to switch to?
> 
> We use git submodules in order to include the jina-hub code into the jina repository. It might have happened, that you (by accident or since you worked on an older master branch) changed to an older version of this git submodule. In order to get back to the latest one, you can run `git submodule update --init` which should set it back to the currently used one. Please give it a try. Furthermore, here is an example to showcase how changing the head of the submodule would look like in the jina repository:
> 
> image

@maximilianwerk I did try running 'git submodule update --init' now. But there are no changes observed upon running the command.

**Comment by maximilianwerk** [comment]
Could you paste me the results from the `git status` command?

**Comment by rameshwara** [comment]
> Could you paste me the results from the `git status` command?

image

**Comment by maximilianwerk** [comment]
> > Could you paste me the results from the `git status` command?
> 
> image

I am still not sure, what went wrong, but I see that everything seems to work in the other PR. Great!

**Pull request closed without merging**